precompiles/staking/method_distribute.go

```diff
@@ -0,0 +1,116 @@
+package staking
+
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+
+	ptypes "github.com/zeta-chain/node/precompiles/types"
+	fungibletypes "github.com/zeta-chain/node/x/fungible/types"
+)
+
+var (
+	zrc20lockerAddress = common.HexToAddress("0x0000000000000000000000000000000000000067")
+)
+
+// function distribute(address zrc20, uint256 amount) external returns (bool success)
+func (c *Contract) distribute(
+	ctx sdk.Context,
+	evm *vm.EVM,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != 2 {
+		return nil, &(ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		})
+	}
+
+	// Unpack arguments and check if they are valid.
+	zrc20Addr, amount, err := unpackDistributeArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the original caller address. Necessary for LockZRC20 to work.
+	caller, err := ptypes.GetEVMCallerAddress(evm, contract)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the coinSet in advance, if this step fails do not lock ZRC20.
+	coinSet, err := ptypes.CreateCoinSet(ptypes.ZRC20ToCosmosDenom(zrc20Addr), amount)
+	if err != nil {
+		return nil, err
+	}
+
+	// LockZRC20 locks the ZRC20 under the locker address.
+	// It performs all the necessary checks such as allowance in order to execute a transferFrom.
+	// - spender is the staking contract address (c.Address()).
+	// - owner is the caller address.
+	// - locker is the bank address. Assets are locked under this address to prevent liquidity fragmentation.
+	if err := c.fungibleKeeper.LockZRC20(ctx, c.zrc20ABI, zrc20Addr, c.Address(), caller, zrc20lockerAddress, amount); err != nil {
+		return nil, &ptypes.ErrUnexpected{
+			When: "LockZRC20InBank",
+			Got:  err.Error(),
+		}
+	}
+
+	// With the ZRC20 locked, proceed to mint the cosmos coins.
+	if err := c.bankKeeper.MintCoins(ctx, fungibletypes.ModuleName, coinSet); err != nil {
+		return nil, &ptypes.ErrUnexpected{
+			When: "MintCoins",
+			Got:  err.Error(),
+		}
+	}
+
+	// Send the coins to the FeePool.
+	if err := c.bankKeeper.SendCoinsFromModuleToModule(ctx, fungibletypes.ModuleName, authtypes.FeeCollectorName, coinSet); err != nil {
```

Comment:
What happens if this step fails ? Are the funds still locked or the whole tx is discarded ( like a regular cosmos msg )

Reply:
I think it's like regular Cosmos message but need to confirm (should be asserted in unit tests)

Reply:
@fbac do we have an assertion for this?

Reply:
Yes, this is covered in the e2e tests. We check explicitly for the FeeCollector balance before and after sending coins.
Do we think we'd need extra assertions?